designs/credentialManagement/design.md

```diff
@@ -0,0 +1,115 @@
+intuitive abstraction layer on top of some of these sources so that the implementation details are hidden and a 
+consistent user experience can be provided, while also enabling future expansion of new sources, including possibly by 
+third party plugins.
+
+## Specification
+
+This toolkit will introduce the following concepts:
+
+1. `ToolkitCredentialsProvider` - An abstract class that implements [AwsCredentialProvider] in the SDK. This class does not
+do and actual resolving of credentials, but instead leaves that that to concrete implementations. It instead provides
+an ID that is globally unique across all credential providers as well as defining a way to generate a display name.
+
+2. `ToolkitCredentialsProviderFactory` - Factory interface that knows how to create one or more `ToolkitCredentialsProvider`
+for a credential source. I.e. The `ProfileToolkitCredentialsProviderFactory` will create a `ProfileToolkitCredentialsProviderFactory`
+for each valid profile in the shared credentials file.
+
+3. `ToolkitCredentialsProviderRegistry` - This interface acts as a holder for all `ToolkitCredentialsProviderFactory`. Concrete 
+implementation uses IntelliJ extension system to load and store factories from `plugin.xml` instead of managing a hard 
+coded list. This enables the system to be extended in a dynamic manor by other plugins.
+
+4. `ToolkitCredentialsProviderManager` - This class acts as the union of all `ToolkitCredentialsProviderFactory`. Its 
+job is to be able to list all `ToolkitCredentialsProvider` and return a one when referenced by its global ID. It also
+has the ability to have listeners registered to it so they can listen for changes when `ToolkitCredentialsProvider` are
+added in removed such as when the shared credentials file is modified.
+
+5. `Active Credentials` - Represents the current credentials that the toolkit uses to perform actions or defaults to when
+more than one option is possible. Due to the nature of the IntellJ IDE being multiple windows in one JVM, each project
+window can have a different active credential selected.
+
+### Extension System
+
+We register all `ToolkitCredentialsProviderFactory` through the IntelliJ extension point system by creating a custom
+extension point bean (`CredentialProviderFactoryEP`) under the FQN `aws.toolkit.credentialProviderFactory` which has a 
+single `implementation` attribute. The extension system is only queried once, and the list of instances of
+`ToolkitCredentialsProviderFactory` must be immutable.
+
+Example of usage:
+```xml
+    <extensions defaultExtensionNs="aws.toolkit">
+        <credentialProviderFactory implementation="software.aws.toolkits.jetbrains.core.credentials.profiles.ProfileCredentialProviderFactory"/>
+    </extensions>
+```
+
+### Credential Validation
+
+In order to validate the credentials returned by the  `ToolkitCredentialsProvider` are valid, we make a call to 
```

Comment:
"to validate that x are valid" is redundant. Consider one of:
* "In order to verify the credentials returned by the `ToolkitCredentialsProvider` are valid,"
* "In order to validate the credentials returned by the `ToolkitCredentialsProvider`,"